devices/src/virtio/block.rs

```diff
@@ -1055,7 +1055,7 @@ mod tests {
             0,
             EpollHandlerPayload::Empty,
         );
-        assert!(r.is_err);
+        assert!(r.is_err());
```

Comment:
There are multiple error types returned now. `assert!(r.is_err)` is not enough. Please write comprehensive unit tests that trigger all possible error cases and `match` on each one of them respectively.
Note that `assert_eq` won't work out of the box unless you `derive(Clone)` on the error types; if that's not possible because of nested errors, you can `match` on the string obtained via `format!("{:?}", e)` as `Debug` is already `derive`d.

Reply:
Yeah, good call. I'm trying to add more exhaustive unit tests, but it's unclear to me (a noob), how to make it so that I can trigger the error case for certain things like`FailedReadingQueue`.

In other words, I need to make `self.queue_evt.read()` fail, but I'm not sure how to best do it

https://github.com/firecracker-microvm/firecracker/blob/19dec14be93795bcde0abc70729db3202d01bbd1/devices/src/virtio/block.rs#L357-L369

Any advice would be greatly appreciated :)

Reply:
Oops, I answered somewhere else already, re-answering where I should have :slightly_smiling_face: 
You can force `read` to fail by closing the eventfd **in the test** before calling it:
```diff
diff --git a/devices/src/virtio/block.rs b/devices/src/virtio/block.rs
index 3bd3651..98ffdcb 100644
--- a/devices/src/virtio/block.rs
+++ b/devices/src/virtio/block.rs
@@ -1064,6 +1064,7 @@ mod tests {
         let m = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
         let (mut h, _vq) = default_test_blockepollhandler(&m);
 
+        unsafe { libc::close(h.queue_evt.as_raw_fd()) };
         let r = h.handle_event(
             QUEUE_AVAIL_EVENT as DeviceEventT,
             0,

```

Reply:
Awesome! Thanks for the pointer 😄 I’ll run with this.

Reply:
Don't do that, see my other comment in this PR.

Reply:
An example of a safe (but arguably ugly) way to do that is to introduce "`cfg(test)`-only internal state mutators". These can be `cfg(test)`-only members of the resource that alter it's behavior during tests.

An example of this can be found in https://github.com/firecracker-microvm/firecracker/commit/620782b03d0684656036b60721efa9d4f0c4ce32

Though my opinion is we might not want that for the EventFd crate since it will make it harder to expose it on crates.io

Reply:
Ok, I've left out the tests that check for errors coming from FD.s